# **Homework**: Data talks club data engineering zoomcamp Data loading workshop

Hello folks, let's practice what we learned - Loading data with the best practices of data engineering.

Here are the exercises we will do




# 1. Use a generator

Remember the concept of generator? Let's practice using them to futher our understanding of how they work.

Let's define a generator and then run it as practice.

**Answer the following questions:**

- **Question 1: What is the sum of the outputs of the generator for limit = 5?**
- **Question 2: What is the 13th number yielded**

I suggest practicing these questions without GPT as the purpose is to further your learning.

In [ ]:
!pip install dlt[duckdb]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 540.2/540.2 kB 8.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 196.4/196.4 kB 23.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 138.7/138.7 kB 15.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 384.9/384.9 kB 36.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 137.9/137.9 kB 18.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 346.6/346.6 kB 30.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.6/49.6 kB 5.6 MB/s eta 0:00:00


In [ ]:
5 ** 0.5

2.23606797749979

# 2. Append a generator to a table with existing data


Below you have 2 generators. You will be tasked to load them to duckdb and answer some questions from the data

1. Load the first generator and calculate the sum of ages of all people. Make sure to only load it once.
2. Append the second generator to the same table as the first.
3. **After correctly appending the data, calculate the sum of all ages of people.**




# 3. Merge a generator

Re-use the generators from Exercise 2.

A table's primary key needs to be created from the start, so load your data to a new table with primary key ID.

Load your first generator first, and then load the second one with merge. Since they have overlapping IDs, some of the records from the first load should be replaced by the ones from the second load.

After loading, you should have a total of 8 records, and ID 3 should have age 33.

Question: **Calculate the sum of ages of all the people loaded as described above.**


# Solution: First make sure that the following modules are installed:

In [5]:
#Install the dependencies
%%capture
!pip install dlt[duckdb]

In [ ]:
# to do: homework :)

# Question 1: What is the sum of the outputs of the generator for limit = 5?

In [ ]:
def square_root_generator(limit):
    n = 1
    while n <= limit:
        yield n ** 0.5
        n += 1

# Example usage:
limit = 5
generator = square_root_generator(limit)
suma = 0

for sqrt_value in generator:
    suma = sqrt_value + suma
    print(sqrt_value)

print (suma)



1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
8.382332347441762


## Respuesta :  8.382332347441762

# Question 2: What is the 13th number yielded


In [1]:
def square_root_generator(limit):
    n = 1
    while n <= limit:
        yield n ** 0.5
        n += 1

# Example usage:
limit = 13
generator = square_root_generator(limit)

for sqrt_value in generator:
    print(sqrt_value)



1.0
1.4142135623730951
1.7320508075688772
2.0
2.23606797749979
2.449489742783178
2.6457513110645907
2.8284271247461903
3.0
3.1622776601683795
3.3166247903554
3.4641016151377544
3.605551275463989


# 1. Load the first generator and calculate the sum of ages of all people. Make sure to only load it once.

In [6]:

def people_1():
    for i in range(1, 6):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 25 + i, "City": "City_A"}

for person in people_1():
    print(person)


def people_2():
    for i in range(3, 9):
        yield {"ID": i, "Name": f"Person_{i}", "Age": 30 + i, "City": "City_B", "Occupation": f"Job_{i}"}


for person in people_2():
    print(person)


{'ID': 1, 'Name': 'Person_1', 'Age': 26, 'City': 'City_A'}
{'ID': 2, 'Name': 'Person_2', 'Age': 27, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 28, 'City': 'City_A'}
{'ID': 4, 'Name': 'Person_4', 'Age': 29, 'City': 'City_A'}
{'ID': 5, 'Name': 'Person_5', 'Age': 30, 'City': 'City_A'}
{'ID': 3, 'Name': 'Person_3', 'Age': 33, 'City': 'City_B', 'Occupation': 'Job_3'}
{'ID': 4, 'Name': 'Person_4', 'Age': 34, 'City': 'City_B', 'Occupation': 'Job_4'}
{'ID': 5, 'Name': 'Person_5', 'Age': 35, 'City': 'City_B', 'Occupation': 'Job_5'}
{'ID': 6, 'Name': 'Person_6', 'Age': 36, 'City': 'City_B', 'Occupation': 'Job_6'}
{'ID': 7, 'Name': 'Person_7', 'Age': 37, 'City': 'City_B', 'Occupation': 'Job_7'}
{'ID': 8, 'Name': 'Person_8', 'Age': 38, 'City': 'City_B', 'Occupation': 'Job_8'}


In [7]:
import dlt

# define the connection to load to.
# We now use duckdb, but you can switch to Bigquery later
generators_pipeline = dlt.pipeline(destination='duckdb', dataset_name='generators')


# we can load any generator to a table at the pipeline destnation as follows:
info = generators_pipeline.run(people_1,
										table_name="people_1",
										write_disposition="replace")

# the outcome metadata is returned by the load and we can inspect it by printing it.
print(info)

# we can load the next generator to the same or to a different table.

info = generators_pipeline.run(people_2,
										table_name="people_2",
										write_disposition="replace")

print(info)

Pipeline dlt_colab_kernel_launcher load step completed in 0.38 seconds
1 load package(s) were loaded to destination duckdb and into dataset generators
The duckdb destination used duckdb:////content/dlt_colab_kernel_launcher.duckdb location to store data
Load package 1707777081.9711418 is LOADED and contains no failed jobs
Pipeline dlt_colab_kernel_launcher load step completed in 0.26 seconds
1 load package(s) were loaded to destination duckdb and into dataset generators
The duckdb destination used duckdb:////content/dlt_colab_kernel_launcher.duckdb location to store data
Load package 1707777082.7813208 is LOADED and contains no failed jobs


In [8]:
# show outcome

import duckdb

conn = duckdb.connect(f"{generators_pipeline.pipeline_name}.duckdb")

# let's see the tables
conn.sql(f"SET search_path = '{generators_pipeline.dataset_name}'")
print('Loaded tables: ')
display(conn.sql("show tables"))

# and the data

print("\n\n\n people_1 table below:")

people_1 = conn.sql("SELECT * FROM people_1").df()
display(people_1)

print("\n\n\n suma de edades de people_1:")

people_1_age = conn.sql("SELECT sum(Age) as Age_suma FROM people_1").df()
display(people_1_age)


Loaded tables: 


┌─────────────────────┐
│        name         │
│       varchar       │
├─────────────────────┤
│ _dlt_loads          │
│ _dlt_pipeline_state │
│ _dlt_version        │
│ people_1            │
│ people_2            │
└─────────────────────┘




 people_1 table below:


id      name  age    city        _dlt_load_id         _dlt_id
0   1  Person_1   26  City_A  1707777081.9711418  +8VAHtBnBC57pg
1   2  Person_2   27  City_A  1707777081.9711418  SJZl58sPeyHnhA
2   3  Person_3   28  City_A  1707777081.9711418  x7p9c3UFgtY+MQ
3   4  Person_4   29  City_A  1707777081.9711418  Uf9m8vS8lCtBwQ
4   5  Person_5   30  City_A  1707777081.9711418  DdYkZHF96G6RgQ




 suma de edades de people_1:


Age_suma
0     140.0

# Question 3. Append the 2 generators. After correctly appending the data, calculate the sum of all ages of people.

In [26]:

print("\n\n\n edades_total table below:")

edades_total = conn.sql("SELECT id, name, age FROM people_1 UNION ALL SELECT id, name, age FROM people_2").df()
display(edades_total)

print("\n\n\n El total de las edades de People1 y People2 es:")

edades_total_suma = conn.sql("SELECT sum(Age) as Age_suma FROM edades_total").df()
display(edades_total_suma)






 edades_total table below:


id      name  age
0    1  Person_1   26
1    2  Person_2   27
2    3  Person_3   28
3    4  Person_4   29
4    5  Person_5   30
5    3  Person_3   33
6    4  Person_4   34
7    5  Person_5   35
8    6  Person_6   36
9    7  Person_7   37
10   8  Person_8   38




 El total de las edades de People1 y People2 es:


Age_suma
0     353.0

## Respuesta 353

# Question 4. Merge the 2 generators using the ID column. Calculate the sum of ages of all the people loaded as described above.

In [31]:


print("\n\n\n people_1 append people_2 table below:")

append_table = conn.sql("SELECT p1.id, p1.name, p1.age FROM people_1 AS p1 LEFT OUTER JOIN people_2 AS p2 ON p1.id = p2.id").df()
display(append_table)


# As you can see, the same data was loaded in both cases.







 people_1 append people_2 table below:


id      name  age
0   3  Person_3   28
1   4  Person_4   29
2   5  Person_5   30
3   1  Person_1   26
4   2  Person_2   27

# 3. Merge a generator

Re-use the generators from Exercise 2.

A table's primary key needs to be created from the start, so load your data to a new table with primary key ID.

Load your first generator first, and then load the second one with merge. Since they have overlapping IDs, some of the records from the first load should be replaced by the ones from the second load.

After loading, you should have a total of 8 records, and ID 3 should have age 33.

Question: **Calculate the sum of ages of all the people loaded as described above.**

In [ ]:
info = generators_pipeline.run(people_1,
										table_name="people_1_pk",
										write_disposition="replace",
                    primary_key="id")

info = generators_pipeline.run(people_2,
										table_name="people_2_pk",
										write_disposition="replace",
                    primary_key="id")

conn = duckdb.connect(f"{generators_pipeline.pipeline_name}.duckdb")

# let's see the tables
conn.sql(f"SET search_path = '{generators_pipeline.dataset_name}'")
print('Loaded tables: ')
display(conn.sql("show tables"))

print("\n\n\n people_1_pk")
people_1_pk = conn.sql("SELECT * FROM people_1_pk").df()
display(people_1_pk)

print("\n\n\n people_2_pk")
people_2_pk = conn.sql("SELECT * FROM people_2_pk").df()
display(people_2_pk)

# and the data

print("\n\n\n merge")
merge_table = conn.sql("SELECT id,name,age,city	 FROM people_1_pk as p1 inner JOIN people_2_pk as p2 on p1.id = p2.id").df()
display(merge_table)


Loaded tables: 


┌─────────────────────┐
│        name         │
│       varchar       │
├─────────────────────┤
│ _dlt_loads          │
│ _dlt_pipeline_state │
│ _dlt_version        │
│ people_1            │
│ people_1_pk         │
│ people_2            │
│ people_2_pk         │
└─────────────────────┘




 people_1_pk


id      name  age    city        _dlt_load_id         _dlt_id
0   1  Person_1   26  City_A  1707355344.4441853  Xi0IPfANJNCXEg
1   2  Person_2   27  City_A  1707355344.4441853  40zRYGKp2SriRQ
2   3  Person_3   28  City_A  1707355344.4441853  6rnjDUwGIwkUbA
3   4  Person_4   29  City_A  1707355344.4441853  olm+beeBmZJAUg
4   5  Person_5   30  City_A  1707355344.4441853  vn14y1uqcTX3Iw




 people_2_pk


id      name  age    city occupation        _dlt_load_id         _dlt_id
0   3  Person_3   33  City_B      Job_3  1707355345.1895993  sSY8vbd8czJSaw
1   4  Person_4   34  City_B      Job_4  1707355345.1895993  YofeY6LfyoEDXg
2   5  Person_5   35  City_B      Job_5  1707355345.1895993  tXIV8Giumi66Cg
3   6  Person_6   36  City_B      Job_6  1707355345.1895993  MHcxZyn2iznOOQ
4   7  Person_7   37  City_B      Job_7  1707355345.1895993  gueySisaJ1Fyog
5   8  Person_8   38  City_B      Job_8  1707355345.1895993  wJPdaNkyZJ0YIw




 merge


id      name  age    city        _dlt_load_id         _dlt_id  id_2  \
0   3  Person_3   28  City_A  1707355344.4441853  6rnjDUwGIwkUbA     3   
1   4  Person_4   29  City_A  1707355344.4441853  olm+beeBmZJAUg     4   
2   5  Person_5   30  City_A  1707355344.4441853  vn14y1uqcTX3Iw     5   

     name_2  age_2  city_2 occupation      _dlt_load_id_2       _dlt_id_2  
0  Person_3     33  City_B      Job_3  1707355345.1895993  sSY8vbd8czJSaw  
1  Person_4     34  City_B      Job_4  1707355345.1895993  YofeY6LfyoEDXg  
2  Person_5     35  City_B      Job_5  1707355345.1895993  tXIV8Giumi66Cg

In the table methode people_1 we have ID 1 to 5. In the methode people_2 we have ID 3 to 8.
When we want to merge with the primary key of "ID" we would have
ID 1 and 2 from people_1 AND
ID 3 to 8 from people_2.

Questions? difficulties? We are here to help.
- DTC data engineering course channel: https://datatalks-club.slack.com/archives/C01FABYF2RG
- dlt's DTC cohort channel: https://dlthub-community.slack.com/archives/C06GAEX2VNX